packages/mdc-menu-surface/foundation.ts

```diff
@@ -0,0 +1,500 @@
+    const availableRight = isAlignedRight ? viewportDistance.right - this.anchorMargin_.right :
+      viewportDistance.right + anchorSize.width - this.anchorMargin_.left;
+
+    const leftOverflow = surfaceSize.width - availableLeft;
+    const rightOverflow = surfaceSize.width - availableRight;
+
+    if ((leftOverflow < 0 && isAlignedRight && isRtl) ||
+        (avoidHorizontalOverlap && !isAlignedRight && leftOverflow < 0) ||
+        (rightOverflow > 0 && leftOverflow < rightOverflow)) {
+      corner |= CornerBit.RIGHT;
+    }
+
+    return corner;
+  }
+
+  /**
+   * @param corner Origin corner of the menu surface.
+   * @return Maximum height of the menu surface, based on available space. 0 indicates should not be set.
+   */
+  private getMenuSurfaceMaxHeight_(corner: Corner): number {
+    const {viewportDistance} = this.measurements_;
+
+    let maxHeight = 0;
+    const isBottomAligned = Boolean(corner & CornerBit.BOTTOM);
+    const {MARGIN_TO_EDGE} = MDCMenuSurfaceFoundation.numbers;
+
+    // When maximum height is not specified, it is handled from css.
+    if (isBottomAligned) {
+      maxHeight = viewportDistance.top + this.anchorMargin_.top - MARGIN_TO_EDGE;
+      if (!(this.anchorCorner_ & CornerBit.BOTTOM)) {
+        maxHeight += this.measurements_.anchorSize.height;
+      }
+    } else {
+      maxHeight =
+          viewportDistance.bottom - this.anchorMargin_.bottom + this.measurements_.anchorSize.height - MARGIN_TO_EDGE;
+      if (this.anchorCorner_ & CornerBit.BOTTOM) {
+        maxHeight -= this.measurements_.anchorSize.height;
+      }
+    }
+
+    return maxHeight;
+  }
+
+  /**
+   * @param corner Origin corner of the menu surface.
+   * @return Horizontal offset of menu surface origin corner from corresponding anchor corner.
+   */
+  private getHorizontalOriginOffset_(corner: Corner): number {
+    const {anchorSize} = this.measurements_;
+
+    // isRightAligned corresponds to using the 'right' property on the surface.
+    const isRightAligned = Boolean(corner & CornerBit.RIGHT);
+    const avoidHorizontalOverlap = Boolean(this.anchorCorner_ & CornerBit.RIGHT);
+
+    if (isRightAligned) {
+      const rightOffset =
+          avoidHorizontalOverlap ? anchorSize.width - this.anchorMargin_.left : this.anchorMargin_.right;
+
+      // For hoisted or fixed elements, adjust the offset by the difference between viewport width and body width so
+      // when we calculate the right value (`adjustPositionForHoistedElement_`) based on the element position,
+      // the right property is correct.
+      if (this.isHoistedElement_ || this.isFixedPosition_) {
+        return rightOffset - (this.measurements_.viewportSize.width - this.measurements_.bodySize.width);
+      }
+
+      return rightOffset;
+    }
+
+    return avoidHorizontalOverlap ? anchorSize.width - this.anchorMargin_.right : this.anchorMargin_.left;
+  }
+
+  /**
+   * @param corner Origin corner of the menu surface.
```

Comment:
remove closure comments

Reply:
This is actually [TypeDoc](https://typedoc.org/guides/doccomments/) syntax 😀